src/typing/env_js.ml

```diff
@@ -661,24 +771,21 @@ let merge_env =
           (string_of_reason reason)
           name)
       (* global lookups may leave new entries in one or both child envs *)
-      | None, child1, child2 when env0 = [] ->
+      | None, _, _ when env0_ = [] ->
         (* ...in which case we can forget them *)
         ()
-      (* otherwise, non-refinement uneven distributions are asserts. *)
-      (* TODO:
-         Asserting on all uneven distributions is too strict for now.
-         let_env can interact with path-dependent refinement to leave
-         residual entries in child contexts, when catch blocks combine
-         conditionals and early returns.
-         So for now we need to let extra child entries pass, though we
-         should still assert on extra parent entries.
-         Once let/const/LexScope is in place, we can reimplement let_env
-         in terms of those constructs instead, and the full assert can
-         be reinstated.
-       *)
-      | None, _, _ ->
+      (* changeset entry exists only in lex scope *)
+      | None, None, Some e when is_lex scope2 && Entry.is_lex e ->
+        ()
+      | None, Some e, None when is_lex scope1 && Entry.is_lex e ->
         ()
-      | Some _ as orig, child1, child2 ->
+      (* walk through uneven lex scopes *)
```

Comment:
:) ...so I was about to just bang out a show-don't-tell, but first let me just ask: will these lists ever actually be ragged? 
  - they should be uniform from the bottom (global) up to the topmost VarScope, since we only ever merge after traversing local control structures
  - any nonuniform LexScopes should have been popped off before the merge takes place, I think. This is from first principles, but a quick scan of type_inference_js.ml in this PR seems to confirm it...

If you can set me straight with an example that makes it ragged, please do. :) Either way, I'll come back with a scratch version of merge_entry that explains what I'm trying to get at... basically I always try to a) constrain the space of possibilities to only what should happen (for instance, if there's any possibility of raggedness, it definitely stops at the topmost VarScope) and b) partition what's left hierarchically, as soon as the flat partition starts to have that whack-a-mole feel to it.